voucher/views.py

```diff
@@ -1,3 +1,201 @@
+
+    def form_valid(self, form):
+        """
+        If parsing is successful, route based on the voucher
+        """
+        return self.handle_voucher(form.cleaned_data["voucher"])
+
+    def form_invalid(self, form):
+        """
+        If parsing fails, route to email page, else return to form
+        """
+        if parse_error_string in form.errors["voucher"]:
+            return redirect(reverse("voucher:resubmit"))
+        return super().form_invalid(form)
+
+    def get_context_data(self, **kwargs):
+        return {
+            **super().get_context_data(**kwargs),
+            **get_js_settings_context(self.request),
+        }
+
+    def handle_voucher(self, values):
+        """
+        Get or create voucher and route based on current progress (i.e. coupon is redeemed / attached / unattached)
+
+        Args:
+            values (dict): dictionary parsed from voucher PDF
+        """
+        user = self.request.user
+        # Check for an existing voucher
+        old_voucher = Voucher.objects.filter(**values).last()
+        # If a voucher exists, check if it is the same as the uploaded voucher
+        if old_voucher:
+            voucher = old_voucher
+            voucher.uploaded = datetime.now(tz=pytz.UTC)
+            voucher.save()
+        else:
+            Voucher.objects.create(**values, user=user)
+
+        return redirect("voucher:enroll")
+
+
+class EnrollView(LoginRequiredMixin, View):
+    """
+    EnrollView checks the status of the voucher and looks for valid course runs to redeem it for
+
+    On a POST, it redirects to the enrollment URL based on the submitted CouponEligibility object's product and
+    coupon_code
+    """
+
+    def get(self, request):
+        """
+        If voucher is not redeemed and valid coupons exist for course runs matching the input strings,
+        render the enroll form with CouponEligibility objects as choices.
+        """
+        voucher = get_current_voucher(self.request.user)
+        if voucher is None:
+            return redirect("voucher:upload")
+        elif voucher.is_redeemed():
+            return redirect("voucher:redeemed")
+        eligible_choices = match_courses_to_voucher(voucher)
+        if not eligible_choices:
+            return redirect("voucher:resubmit")
+        else:
+            return render(
+                request,
+                "enroll.html",
+                context={
+                    "eligible_choices": eligible_choices,
+                    **get_js_settings_context(self.request),
+                },
+            )
+
+    def post(self, request):
+        """Submit a CouponVersion object and redirect to the enrollment page"""
+        voucher = get_current_voucher(self.request.user)
+        product_id, coupon_id = json.loads(request.POST["coupon_version"])
+        voucher.coupon_id = coupon_id
+        voucher.save()
+        enroll_url = (
+            f"/enroll?product_id={product_id}&code={voucher.coupon.coupon_code}"
```

Comment:
Should be "/checkout?" instead of "/enroll", if you rebase with master that should bring you to the checkout page with the product in the basket and the coupon applied.

Reply:
also `product=<product_id>` instead of `product_id`